### PR TITLE
Make CLI automatically switch from aligned to vertical format

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

Make CLI automatically switch from aligned to vertical format when the output wouldn't fit the current terminal's width. Currently, users can terminate the query with `\G`, but this has a few drawbacks:
* users need to run the query twice if they notice it's too wide
* a table that's too wide will be opened in the pager and won't remain on the screen when writing a follow-up query - users might want to refer to the values from the first query, so they'd need to interrupt writing the second one
* users have to notice the output is too wide - check out the output of `SHOW FUNCTIONS` where the second column is so wide you might not notice there are more
* it's kinda hard to type - backslash is unusual and `G` has to be upper-case

This PR introduces a new `AUTO` output format that automatically switches from `ALIGNED` to `VERTICAL` if the aligned table is wider than the current terminal. But it doesn't make the auto format the default, for backward compatibility. Users need to enable it by using `--output-format-interactive=AUTO` or by adding it to the config file.

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

cli

> How would you describe this change to a non-technical end user or system administrator?

Add a new `AUTO` output format to the CLI client, that switches from `ALIGNED` to `VERTICAL` if the output wouldn't fit the current terminal. 

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# CLI
* Add a new `AUTO` output format, that switches from `ALIGNED` to `VERTICAL` if the output wouldn't fit the current terminal.  ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by nineinchnick** [comment]
This is the equivalent of `\x auto` in `psql`. Setting the env var is the equivalent of adding `\x auto` to `~/.psqlrc`.

**Review by hashhar — COMMENTED** [review]
The feature and implementation seems good to me.

The env var seems overkill to me. What use-case did you have in mind?

**Comment by nineinchnick** [comment]
> The feature and implementation seems good to me.
> 
> The env var seems overkill to me. What use-case did you have in mind?

It's required so that users (well, I mostly ;-) can have it turned on by default, without changing the current default output (aligned). There's no RC file like in PostgreSQL. Without this, it defeats the purpose of it being automatic if I'd have to _remember_ to turn it on on every invocation.

**Comment by hashhar** [comment]
Let's discuss about having a config file then, it seems like a good change IMO since it'll make it much easier to support things like not having to enter lengthy arguments for Kerberos or self-signed tls auth etc. as well instead of just being specific to this single feature.

cc: @electrum @findepi WDYT about a config file to read some defaults for configs from? Would make using the CLI a lot easier.

**Comment by nineinchnick** [comment]
@hashhar would it be enough to read a properties file from the user's home dir? https://picocli.info/#_propertiesdefaultprovider

**Comment by nineinchnick** [comment]
@hashhar I removed the new env var for output format but changing the default in the config file means it'll apply to both interactive and batch modes. I'm thinking about adding yet another flag for interactive mode, but `--output-format-interactive` is a bit too long. Maybe a boolean `--auto-expand` to be used instead of adding a new output format?

**Comment by nineinchnick** [comment]
@hashhar @electrum @findepi is is ready for review.

File: client/trino-cli/src/main/java/io/trino/cli/Console.java [hunk 1]
```diff
@@ -71,14 +69,16 @@
...
 
 @Command(
         name = "trino",
         header = "Trino command line interface",
         synopsisHeading = "%nUSAGE:%n%n",
         optionListHeading = "%nOPTIONS:%n",
         usageHelpAutoWidth = true,
-        versionProvider = VersionProvider.class)
+        versionProvider = VersionProvider.class,
+        defaultValueProvider = PropertiesDefaultProvider.class)
```
Comment (nineinchnick) [inline client/trino-cli/src/main/java/io/trino/cli/Console.java:81]: I somehow _totally_ missed there's already https://github.com/trinodb/trino/blob/master/client/trino-cli/src/main/java/io/trino/cli/Trino.java#L55

I'll remove this.

**Comment by mosabua** [comment]
Should I review for docs yet @nineinchnick ?

**Comment by nineinchnick** [comment]
> Should I review for docs yet @nineinchnick ?

@mosabua yes please.

File: docs/src/main/sphinx/installation/cli.rst [hunk 2]
```diff
@@ -540,6 +532,49 @@ and displays an error message (which is unaffected by the output format):
     Query 20200707_170726_00030_2iup9 failed: line 1:25: Column 'region' cannot be resolved
     SELECT nationkey, name, region FROM tpch.sf1.nation LIMIT 3
 
+.. _cli-output-format:
+
+Output formats
+--------------
+
+The Trino CLI provides the options ``--output-format``
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:540]: Why the weird wrapping?

**Review by mosabua — CHANGES_REQUESTED** [review]

File: docs/src/main/sphinx/installation/cli.rst [hunk 3]
```diff
@@ -405,6 +418,24 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
...
+Configuration file
+------------------
+
+The CLI can read default values for all options from a file. It'll use
+the first one found at these locations, in order:
+* the value of the ``TRINO_CONFIG`` environment variable,
+* ``~/.trino_config``,
+* ``$XDG_CONFIG_HOME/trino/config``.
+
+Use option names as keys, without the ``--`` prefix.
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst:430]: Fully agree, added!

File: docs/src/main/sphinx/installation/cli.rst [hunk 4]
```diff
@@ -405,6 +418,24 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
...
+
+.. warning::
+
+    Avoid putting sensitive information, like the password, in the config file.
+    Trino CLI does not check if the permissions of this file are as strict
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst]: Trino CLI doesn't allow passing the user password as an argument - it either prompts for it or reads it from an environmental variable. There are other options that accept values, like the truststore password, but it's a separate issue to address.

So I'll drop this warning for now.

File: docs/src/main/sphinx/installation/cli.rst [hunk 5]
```diff
@@ -405,6 +418,23 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
...
+Configuration file
+------------------
+
+The CLI can read default values for all options from a file. It uses the first
+file found from the ordered list of locations:
+* File path set as value of the ``TRINO_CONFIG`` environment variable.
+* ``.trino_config`` in the current users home directory.
+* ``$XDG_CONFIG_HOME/trino/config``.
+
+Use option names as keys, without the ``--`` prefix.
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:430]: Maybe reword to something like

You can use all supported options without the -- prefix in the configuration properties file, as displayed in the following example:

File: docs/src/main/sphinx/installation/cli.rst [hunk 6]
```diff
@@ -405,6 +418,23 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
...
+
+.. code-block:: text
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:432]: properties

File: docs/src/main/sphinx/installation/cli.rst (hunk 2 above)
Comment (mosabua) [reply docs/src/main/sphinx/installation/cli.rst:539]: I meant on the whole paragraph .. somehow looks off to me .. anyway .. no worries .. I am just used to rewrap extension usage in vscode..

File: docs/src/main/sphinx/installation/cli.rst (hunk 5 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst]: Thanks!

File: docs/src/main/sphinx/installation/cli.rst [hunk 7]
```diff
@@ -405,6 +415,24 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
 buffer. Press :kbd:`Ctrl+→` (:kbd:`Option+→` on Mac) to accept only the next
 keyword. Continue typing to reject the suggestion.
 
+Configuration file
```
Comment (wendigo) [inline docs/src/main/sphinx/installation/cli.rst:421]: Thanks for documenting this feature that I've added a long time ago :)

File: docs/src/main/sphinx/installation/cli.rst [hunk 8]
```diff
@@ -405,6 +418,24 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
...
+------------------
+
+The CLI can read default values for all options from a file. It uses the first
+file found from the ordered list of locations:
+* File path set as value of the ``TRINO_CONFIG`` environment variable.
```
Comment (mosabua) [inline docs/src/main/sphinx/installation/cli.rst:426]: empty line before list

**Review by mosabua — APPROVED** [review]
Docs look good now.

File: docs/src/main/sphinx/installation/cli.rst (hunk 8 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst:427]: Done!

File: docs/src/main/sphinx/installation/cli.rst (hunk 7 above)
Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst:421]: You can return the favor by reviewing this PR :D

File: docs/src/main/sphinx/installation/cli.rst [hunk 9]
```diff
@@ -405,6 +418,25 @@ Press :kbd:`→` to accept the suggestion and replace the current command line
 buffer. Press :kbd:`Ctrl+→` (:kbd:`Option+→` on Mac) to accept only the next
 keyword. Continue typing to reject the suggestion.
 
+Configuration file
+------------------
+
+The CLI can read default values for all options from a file. It uses the first
+file found from the ordered list of locations:
```
Comment (aakashnand) [inline docs/src/main/sphinx/installation/cli.rst:425]: @nineinchnick What is the meaning of first file in ordered list of locations? Can we specify multiple files? If yes, how do we use multiple files? 

I am not sure if this can be helpful but when specifying boolean properties it might be helpful to mention in the doc the boolean properties can be written like 

`krb5-disable-remote-service-hostname-canonicalization=true`


Additional question, if I have multiple authentication enabled like Kerberos, Password, Can I use multiple files with different list of arguments and how do I activate them?

Comment (nineinchnick) [reply docs/src/main/sphinx/installation/cli.rst:425]: Excellent comment, thanks! I extracted this change to [a separate PR](https://github.com/trinodb/trino/pull/12815) and applied these suggestions, that is explained how to use a custom config and described boolean values.

**Comment by nineinchnick** [comment]
@dain maybe you'd like to review this one?

**Comment by mosabua** [comment]
Could we get this reviewed by whoever necessary and merged?

**Comment by hashhar** [comment]
@electrum can you PTAL?

**Comment by mosabua** [comment]
I think this is a good feature to add.

**Comment by nineinchnick** [comment]
@kokosing WDYT?

File: client/trino-cli/src/main/java/io/trino/cli/Console.java [hunk 10]
```diff
@@ -191,7 +189,18 @@ public boolean run()
                         clientOptions.progress.orElse(false));
             }
 
-            runConsole(queryRunner, exiting, clientOptions.editingMode, clientOptions.progress.orElse(true), clientOptions.disableAutoSuggestion);
+            Optional<String> pager = clientOptions.pager;
+            if (!pager.isPresent()) {
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/Console.java:193]: isEmpty()

File: client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java [hunk 11]
```diff
@@ -148,6 +151,9 @@
     @Option(names = "--output-format", paramLabel = "<format>", defaultValue = "CSV", description = "Output format for batch mode [${COMPLETION-CANDIDATES}] " + DEFAULT_VALUE)
     public OutputFormat outputFormat;
 
+    @Option(names = "--pager", paramLabel = "<pager>", description = "Path to the pager program used to display the query results")
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:157]: Can we provide default value like in case of historyFile?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.List;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class AutoTablePrinter
+        implements OutputPrinter
+{
+    private final int maxWidth;
+    private OutputPrinter delegate;
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:30]: final

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    @Override
+    public void printRows(List<List<?>> rows, boolean complete)
+            throws IOException
+    {
+        delegate.printRows(rows, complete);
+        if (!delegate.equals(fallback)) {
+            if (bufferWriter.toString().indexOf("\n") > maxWidth) {
+                delegate = fallback;
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:64]: I don't like overriding variables.  Why not to call simply `falback` below?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        flush();
+    }
+
+    @Override
+    public void printRows(List<List<?>> rows, boolean complete)
+            throws IOException
+    {
+        delegate.printRows(rows, complete);
+        if (!delegate.equals(fallback)) {
+            if (bufferWriter.toString().indexOf("\n") > maxWidth) {
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:63]: One one hand this class is generic as it takes generic delegate and fallback, but then condition is hardcoded.

Maybe you should have `ConditionalTablePrinter` that would be use create a variable `autoTablePrinter`? WDYT?

```
ConditionalTablePrinter autoTablePrinter = new ConditionalTablePrinter(bufferWriter -> bufferWriter.toString().indexOf("\n") < maxWidth, new AlignedTablePrinter(...), new VerticalRecordPrinter(...));
```?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    {
+        delegate.finish();
+        flush();
+    }
+
+    @Override
+    public void printRows(List<List<?>> rows, boolean complete)
+            throws IOException
+    {
+        delegate.printRows(rows, complete);
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:61]: does it write all rows or just one?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 14 above)
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:63]: Do we have to populate entire results to execute the condition? In case of of huge results it could be slow, right?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+                delegate = fallback;
+                bufferWriter.getBuffer().setLength(0);
+                delegate.printRows(rows, complete);
+            }
+        }
+
+        flush();
+    }
+
+    private void flush()
```
Comment (kokosing) [inline client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:73]: inline?

**Review by kokosing — COMMENTED** [review]
I like this change. I think it is also safe to merge as it does not change any existing functionality, but adds new one.

File: client/trino-cli/src/main/java/io/trino/cli/Console.java (hunk 10 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/Console.java:193]: The CLI is built for Java 8 and `isEmpty` is not available.

File: client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java (hunk 11 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:157]: The default is `less -FXRSn` and because it has arguments, it'll get wrapped in `/bin/sh -c "less -FXRSn"`. That can have some side effects, and I don't want to address this in this PR. Should I extract this commit into a separate PR?

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 14 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:63]: It doesn't take `delegate` and `fallback`, it sets them internally. IMO, making it too generic will make it harder to understand what it does and why is it generic.

Also, in the example you gave, we'll need to pull out the buffered writer out of this class, which breaks the encapsulation. I'd have to wrap this up in a `AutoTablePrinter` class anyway, so I don't see any gain.

Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:63]: There's a buffer of 10k rows higher in the call stack, so it shouldn't be an issue, other than seeing the output format switch between aligned and vertical if the width varies a lot. Seems like a very edge case, considering this has to be explicitly enabled and is intended only for interactive use, not dumping large results into files.

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 13 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:64]: This is to make sure we call `finish()` on the right printer.

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 16 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:73]: Why? It's used in two places.

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 12 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:30]: It's not final.

File: client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java (hunk 15 above)
Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/AutoTablePrinter.java:61]: It writes all of them, but there's a maximum of 10k rows buffered in that list.

File: docs/src/main/sphinx/client/cli.rst [hunk 17]
```diff
@@ -161,10 +161,19 @@ mode:
       EMACS editors. Defaults to ``EMACS``.
   * - ``--http-proxy``
     - Configures the URL of the HTTP proxy to connect to Trino.
+  * - ``--history-file``
+    - Path to the :ref:`history file <cli-history>`. Defaults to ``~/.trino_history``.
   * - ``--network-logging``
     - Configures the level of detail provided for network logging of the CLI.
       Defaults to ``NONE``, other options are ``BASIC``, ``HEADERS``, or
       ``BODY``.
+  * - ``--output-format-interactive=<format>``
```
Comment (nineinchnick) [inline docs/src/main/sphinx/client/cli.rst:170]: It might be a good idea to add this to the example in the `Configuration file` section since setting it to `AUTO` is useful.

File: client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java (hunk 11 above)
Comment (kokosing) [reply client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:157]: Please do.

Comment (nineinchnick) [reply client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java:157]: https://github.com/trinodb/trino/pull/16151

**Comment by nineinchnick** [comment]
@kokosing this is ready to go, PTAL.

**Comment by aakashnand** [comment]
@nineinchnick I also think this is useful. I have faced this problem where there are too many columns, until now I have been using `\G` but this flag now looks promising. Good Job 👍🏼

**Review by kokosing — APPROVED** [review]

**Comment by kokosing** [comment]
Thanks
